### PR TITLE
Add negate kernels (#4488)

### LINKED ISSUE #4488: Interval Negation Incorrect
**Describe the bug**

The negate kernels are incorrect for interval types as they fail to take into account that the values are composite

**To Reproduce**

```
#[test]
fn test_negate() {
    let a =
        IntervalMonthDayNanoArray::from(vec![IntervalMonthDayNanoType::make_value(
            1, -5, 2,
        )]);
    let negated = negate(&a).unwrap();
    assert_eq!(
        negated.value(0),
        IntervalMonthDayNanoType::make_value(-1, 5, -2)
    );
}
```

**Expected behavior**

**Additional context**

We should probably deprecate the generic kernels and implement new kernels as part of #1047

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4488

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
Will add interval tests shortly, realise I forgot

**Comment by alamb** [comment]
> Will add interval tests shortly, realise I forgot

Should I wait to review this PR until you wrote those tests? Or did you add them to https://github.com/apache/arrow-rs/pull/4493 ?

File: arrow-arith/src/numeric.rs [hunk 1]
```diff
@@ -74,6 +74,97 @@ pub fn rem(lhs: &dyn Datum, rhs: &dyn Datum) -> Result<ArrayRef, ArrowError> {
...
+}
+
+macro_rules! neg_wrapping {
+    ($t:ty, $a:ident) => {{
+        let array = $a.as_primitive::<$t>().unary::<_, $t>(|x| x.neg_wrapping());
+        Ok(Arc::new(array))
+    }};
+}
+
+/// Perform `!array`, returning an error on overflow
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:93]:
```suggestion
/// Negates each element of  `array`, returning an error on overflow
```

I am not really sure what "Performs `!array`" means

File: arrow-arith/src/numeric.rs [hunk 2]
```diff
@@ -74,6 +74,97 @@ pub fn rem(lhs: &dyn Datum, rhs: &dyn Datum) -> Result<ArrayRef, ArrowError> {
...
+///
+/// Note: negation of unsigned arrays is not supported and will return in an error,
+/// for wrapping unsigned negation consider using [`neg_wrapping()`]
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:96]:
```suggestion
/// for wrapping unsigned negation consider using [`neg_wrapping`]
```

File: arrow-arith/src/numeric.rs [hunk 3]
```diff
@@ -74,6 +74,97 @@ pub fn rem(lhs: &dyn Datum, rhs: &dyn Datum) -> Result<ArrayRef, ArrowError> {
...
+            })?;
+            Ok(Arc::new(a))
+        }
+        t => Err(ArrowError::InvalidArgumentError(format!(
+            "Invalid arithmetic operation: !{t}"
+        ))),
+    }
+}
+
+/// Perform `!array`, wrapping on overflow for [`DataType::is_integer`]
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:160]:
```suggestion
/// Negates each element of  `array` , wrapping on overflow for [`DataType::is_integer`]
```

File: arrow-arith/src/numeric.rs [hunk 4]
```diff
@@ -74,6 +74,97 @@ pub fn rem(lhs: &dyn Datum, rhs: &dyn Datum) -> Result<ArrayRef, ArrowError> {
...
+                .as_primitive::<Decimal256Type>()
+                .try_unary::<_, Decimal256Type, _>(|x| x.neg_checked())?;
+
+            Ok(Arc::new(a.with_precision_and_scale(*p, *s)?))
+        }
+        Duration(Second) => neg_checked!(DurationSecondType, array),
+        Duration(Millisecond) => neg_checked!(DurationMillisecondType, array),
+        Duration(Microsecond) => neg_checked!(DurationMicrosecondType, array),
+        Duration(Nanosecond) => neg_checked!(DurationNanosecondType, array),
+        Interval(YearMonth) => neg_checked!(IntervalYearMonthType, array),
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:128]: I double checked that [YearMonth](https://docs.rs/arrow/latest/arrow/datatypes/enum.IntervalUnit.html#variant.YearMonth) intervals are stored as number of whole intervals and thus don't need to be treated field by field

**Review by alamb — APPROVED** [review]
Thanks @tustvold  -- looks great to me

File: arrow-arith/src/numeric.rs (hunk 2 above)
Comment (tustvold) [reply arrow-arith/src/numeric.rs:96]: This is necessary to avoid ambiguity, will change to a link

Comment (alamb) [reply arrow-arith/src/numeric.rs]: sorry it just looked strange to me

**Pull request merged**